### PR TITLE
[iOS] Ensure the keyboard bottom inset is zero when keyboard dismiss.

### PR DESCRIPTION
**Pull request by luckysmg**
This PR is to ensure the viewport metrics is correct when keyboard dismiss.

Related issue:
https://github.com/flutter/flutter/issues/99951
https://github.com/flutter/flutter/issues/100220

<details>

<summary>Demo code for first issue</summary>

```dart

import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);
  
  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      onTap: (){
        FocusScope.of(context).unfocus();
      },
      child: Scaffold(
        backgroundColor: Colors.white,
        appBar: AppBar(
          title: Text(widget.title),
        ),
        body: Container(
          color: Colors.blue.shade100,
          child: Column(
            children: [
              const SizedBox(height: 48),
              const Padding(
                padding: EdgeInsets.symmetric(horizontal: 16),
                child: Text(
                  '1. Change iOS keyboard to Bitmoji or Slyder keyboard.\n'
                      '2. Tap on the button below.\n'
                      '3. Navigate back to previous route.\n'
                      '4. Check that shaded blue(Scaffold.body) not covering entire screen.\n'
                      '5. Compare the behavior with OS keyboard',
                  style: TextStyle(fontSize: 20),
                ),
              ),
              const TextField(autofocus: true),
              const Spacer(),
              ElevatedButton(
                onPressed: () {
                  FocusScope.of(context).unfocus();
                  Navigator.of(context).push(CupertinoPageRoute(
                    builder: (context) {
                      return const MyHomePage(title: 'Another Page');
                    },
                  ));
                },
                child: const Text('Hello', style: TextStyle(fontSize: 30)),
              )
            ],
          ),
        ),
        bottomNavigationBar: BottomNavigationBar(
          items: [Icons.abc, Icons.train]
              .map((e) => BottomNavigationBarItem(icon: Icon(e), label: 'hi'))
              .toList(),
        ),
      ),
    );
  }
}

```

</details>

**Preview for https://github.com/flutter/flutter/issues/99951**

https://user-images.githubusercontent.com/49340347/158005259-ff33d209-4330-4c54-ad1a-91d192efac4a.MP4

https://user-images.githubusercontent.com/49340347/158005567-dc0fd6ff-44ae-40de-9431-8260c437a9c9.mp4

<br>

</br>

**Preview for https://github.com/flutter/flutter/issues/100220**

https://user-images.githubusercontent.com/49340347/159438702-040a7d48-fd6c-41d7-b973-c2739859307c.mp4

https://user-images.githubusercontent.com/49340347/159438721-5d1591e9-de50-4b2a-ae63-584ec1297c8b.mp4

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by coldstar96** [comment]
Just out of curiosity, why does keyboard with shorter height properly animates without this change?

**Comment by luckysmg** [comment]
@coldstar96 
Reason: When custom keyboard is dismissed ,the given keyboard frame will intersect with screen....
So the targetBottomInset will be bottom * scale instead of 0.

I think maybe there is some issue in custom keyboard app. But make this change in engine side will make it more universal

**Comment by coldstar96** [comment]
It's unfortunate that iOS keyboard notification is not called with consistency. I've been through so may weird behaviors with iOS framework that this is no surprise. Thanks for the explanation!

File: shell/platform/darwin/ios/framework/Source/FlutterViewControllerTest.mm [hunk 1]
```diff
@@ -196,6 +200,30 @@ - (void)testkeyboardWillChangeFrameWillStartKeyboardAnimation {
...
+  FlutterViewController* viewControllerMock = OCMPartialMock(viewController);
+  CGRect keyboardFrame = CGRectMake(0, 0, 0, 0);
+  BOOL isLocal = YES;
+  NSNotification* fakeNotification = [NSNotification
+      notificationWithName:@""
+                    object:nil
+                  userInfo:@{
+                    @"UIKeyboardFrameEndUserInfoKey" : [NSValue valueWithCGRect:keyboardFrame],
+                    @"UIKeyboardAnimationDurationUserInfoKey" : [NSNumber numberWithDouble:0.25],
+                    @"UIKeyboardIsLocalUserInfoKey" : [NSNumber numberWithBool:isLocal]
```
Comment (jmagman) [inline shell/platform/darwin/ios/framework/Source/FlutterViewControllerTest.mm:219]: Use boxed literals
```suggestion
                    @"UIKeyboardAnimationDurationUserInfoKey" : @(0.25),
                    @"UIKeyboardIsLocalUserInfoKey" : @(isLocal)
```

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm [hunk 2]
```diff
@@ -1175,8 +1175,16 @@ - (void)keyboardWillChangeFrame:(NSNotification*)notification {
...
-  // When keyboard hide, the keyboardWillChangeFrame function will be called to update viewport
-  // metrics. So do not call [self updateViewportMetrics] here again.
+  if (self.targetViewInsetBottom != 0) {
+    // Ensure the keyboard will be dismissed. Just like the keyboardWillChangeFrame,
+    // keyboardWillBeHidden is also in an animation block in iOS sdk, so we don't need to set the
+    // animation curve. Related issue: https://github.com/flutter/flutter/issues/99951
+    self.targetViewInsetBottom = 0;
+    NSDictionary* info = [notification userInfo];
+    NSTimeInterval duration =
+        [[info objectForKey:UIKeyboardAnimationDurationUserInfoKey] doubleValue];
```
Comment (jmagman) [inline shell/platform/darwin/ios/framework/Source/FlutterViewController.mm:1185]: Use subscripting.
```suggestion
    NSTimeInterval duration =
        info[UIKeyboardAnimationDurationUserInfoKey].doubleValue;
```

**Review by jmagman — COMMENTED** [review]
@[coldstar96](https://github.com/coldstar96) were you able to confirm this fixes the issue since you can repro?

Very minor nits.
@gaaclarke can you give this a sanity check since you reviewed https://github.com/flutter/engine/pull/29281 ?

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm (hunk 2 above)
Comment (luckysmg) [reply shell/platform/darwin/ios/framework/Source/FlutterViewController.mm:1185]: Emm using subscripting will not compile successfully
``` 
error: property 'doubleValue' not found on object of type 'id _Nullable'
            info[UIKeyboardAnimationDurationUserInfoKey].doubleValue;
                                                         ^
1 error generated.
```

File: shell/platform/darwin/ios/framework/Source/FlutterViewControllerTest.mm (hunk 1 above)
Comment (luckysmg) [reply shell/platform/darwin/ios/framework/Source/FlutterViewControllerTest.mm]: done

**Review by chinmaygarde — COMMENTED** [review]
This looks good to me. @jmagman can you LGTM this?

**Comment by jmagman** [comment]
@cyanglaz is reviewing.

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm [hunk 3]
```diff
@@ -1175,8 +1175,16 @@ - (void)keyboardWillChangeFrame:(NSNotification*)notification {
 }
 
 - (void)keyboardWillBeHidden:(NSNotification*)notification {
-  // When keyboard hide, the keyboardWillChangeFrame function will be called to update viewport
-  // metrics. So do not call [self updateViewportMetrics] here again.
+  if (self.targetViewInsetBottom != 0) {
```
Comment (cyanglaz) [inline shell/platform/darwin/ios/framework/Source/FlutterViewController.mm:1178]: I think we should add the same "isLocal" check in keyboardWillChangeFrame here: 

```
  if (@available(iOS 9, *)) {
    // Ignore keyboard notifications related to other apps.
    id isLocal = info[UIKeyboardIsLocalUserInfoKey];
    if (isLocal && ![isLocal boolValue]) {
      return;
    }
  }
```

**Comment by luckysmg** [comment]
Hi @cyanglaz ,What's your iPhone version ? My iPhone can reproduce it.
iPhone 13 pro. iOS 15.4

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm (hunk 3 above)
Comment (luckysmg) [reply shell/platform/darwin/ios/framework/Source/FlutterViewController.mm:1198]: done

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm [hunk 4]
```diff
@@ -1180,8 +1180,30 @@ - (void)keyboardWillChangeFrame:(NSNotification*)notification {
...
+    id isLocal = info[UIKeyboardIsLocalUserInfoKey];
+    if (isLocal && ![isLocal boolValue]) {
+      return;
+    }
+  }
+
+  // Ignore keyboard notifications if engine’s viewController is not current viewController.
+  if ([_engine.get() viewController] != self) {
+    return;
+  }
```
Comment (cyanglaz) [inline shell/platform/darwin/ios/framework/Source/FlutterViewController.mm:1196]: nits: This code is the same as in keyboardWillChangeFrame, maybe we can refactor it into a method.

**Comment by cyanglaz** [comment]
cc @jmagman or @chinmaygarde for a second LGTM

**Pull request merged**